compose/config/validation.py

```diff
@@ -107,15 +107,73 @@ def _parse_key_from_error_msg(error):
     def _clean_error_message(message):
         return message.replace("u'", "'")
 
-    def _parse_valid_types_from_schema(schema):
+    def _parse_valid_types_from_validator(validator):
         """
-        Our defined types using $ref in the schema require some extra parsing
-        retrieve a helpful type for error message display.
+        A validator value can be either an array of valid types or a string of
+        a valid type. Parse the valid types and prefix with the correct article.
         """
-        if '$ref' in schema:
-            return schema['$ref'].replace("#/definitions/", "").replace("_", " ")
+        pre_msg_type_prefix = "a"
+        last_msg_type_prefix = "a"
+        types_requiring_an = ["array", "object"]
+
+        if isinstance(validator, list):
+            last_type = validator.pop()
+            types_from_validator = ", ".join(validator)
+
+            if validator[0] in types_requiring_an:
+                pre_msg_type_prefix = "an"
+
+            if last_type in types_requiring_an:
+                last_msg_type_prefix = "an"
+
+            msg = "{} {} or {} {}".format(
+                pre_msg_type_prefix,
+                types_from_validator,
+                last_msg_type_prefix,
+                last_type
+            )
         else:
-            return str(schema['type'])
+            if validator in types_requiring_an:
+                pre_msg_type_prefix = "an"
```

Comment:
This check seems to be repeated a few times. I think it would be good to make this a function, which would let us remove all the prefix variables and `types_requiring_an` from this function

```python

def anglicize_validator(validator):
    if validator in ["array", "object"]:
        return 'an ' + validator
    return 'a ' + validator
```

I think this change would return the size of this function quite a bit.